### PR TITLE
[GraalJS] Add support for ES6+ to automation via GraalJS

### PR DESCRIPTION
**Pull request by jpg0**
Signed-off-by: Jonathan Gilbert <jpg@trillica.com>

Bundle to add ES6+ support for automation via GraalJS

OSGi bundle to provide a GraalJS runtime for automation before Graal is bundled with the JDK (e.g. using Java 8 in this case). GraalJS is significantly more capable than Nashorn, which now deprecated.

This is a drop-in replacement for Nashorn, the only incompatibilities potentially being existing scripts which use [Nashorn-specific features](https://github.com/graalvm/graaljs/blob/master/docs/user/NashornMigrationGuide.md) (which will need to be addressed at some point regardless). It does NOT replace the scripting engine for transformations (although this would be relatively straightforward to add).

Note that this bundle also enables CommonJS module support by default (although can be disabled), and therefore allows `npm`-installation of most of the existing available 3rd party JS libraries.

Discussion originally started here: https://community.openhab.org/t/graalvm-for-automation/84116/11

Binary [available](https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.automation.module.script.graaljs/2.5.0-SNAPSHOT/org.openhab.automation.module.script.graaljs-2.5.0-SNAPSHOT.jar) for testing.

### CODE REVIEW

**Comment by davidgraeff** [comment]
This is +83,314 lines of added code. The change for this going through review is low to be honest.
I have checked and you are commiting all  kind of javascript test files of the GraalVM project? If so, maybe reduce to one integration test. The How's and Why's of the engine is already tested in their own repo I guess.

If you have further ideas to cut down on lines please apply those. A regular binding has around 1.5k SLOC.

**Comment by jpg0** [comment]
@davidgraeff you're completely right; apologies for that. The large amount of code was testing for specific edge-cases in loading 3rd party libraries' modules. I've removed it entirely as it's mostly tested elsewhere via mocks. Down to a much more manageable size.

**Comment by kaikreuzer** [comment]
Thanks @jpg0, 2200 LoC look much better than before ;-)

I am wondering, whether this is the right repo for the PR, though. We don't have a add-on category for scripting engines and it probably does not make much sense to introduce such a category.

As you mention, this is rather a replacement for the Nashorn integration, so it should be considered a core component (maybe one that isn't installed by default due to its size, though).

@openhab/core-maintainers What's your view on this? Where should such code best live?

**Comment by jpg0** [comment]
I just realised that somehow an old factory file was left in there which broke classloading order. I've push a new commit to fix this (and also simplified logging).

File: bundles/org.openhab.automation.module.script.graaljs/src/main/java/org/openhab/automation/module/script/graaljs/commonjs/internal/FilesystemFolder.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * File folder implementation
+ *
+ * @author Jonathan Gilbert - Initial contribution
+ */
+@NonNullByDefault
+public class FilesystemFolder extends AbstractFolder {
```
Comment (martinvw) [inline bundles/org.openhab.automation.module.script.graaljs/src/main/java/org/openhab/automation/module/script/graaljs/commonjs/internal/FilesystemFolder.java]: While quickly looking into this PR I noted that the formatting is non-standard here, please format it using Eclipse with the standard formatter.

BTW I'm really happy to see that you got it down to 2200 lines, it makes it reviewable!

**Comment by wborn** [comment]
> What's your view on this? Where should such code best live?

All these scripting engines would be useful for OHC derivatives. But it will have a big impact on the download size if they are all included by default. I see that several of the Maven dependencies are 10MB or more.

File: bundles/org.openhab.automation.module.script.graaljs/pom.xml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    <properties>
+        <bnd.importpackage>
+            !sun.misc.*,
+            !sun.reflect.*,
+            !com.sun.management.*,
+            !jdk.internal.reflect.*,
+            !jdk.vm.ci.services
+        </bnd.importpackage>
+        <graal.version>19.2.0.1</graal.version>
```
[the remark this answers is not in the document]
Comment (jpg0) [reply bundles/org.openhab.automation.module.script.graaljs/pom.xml]: Well spotted. I'll get this version bumped - I know that some of the classloading and language discovery has changed between these versions so I'll need to make sure that it all still works fine in OSGi, but upon initial look it appears simpler (or at least more compatible!).

**Comment by openhab-5iver** [comment]
> But it will have a big impact on the download size if they are all included by default.

I plan to submit a similar PR for a Jython this weekend (:tada:). Once closer to OH3, I plan to open an issue to discuss the possibility of removing the rule engines, scripted automation, script engine factories, module types, rule templates, etc. from OHC and create a new repo for them. Due to this, my suggestion is to put everything automation related into OHC for now. This is where I've built the Jython script engine factory, although it also worked as an addon. Jython is ~40MB.

Once I have the PR in for Jython, I plan to review this PR, which I see several issues with.

File: bundles/org.openhab.automation.module.script.graaljs/src/main/java/org/openhab/automation/module/script/graaljs/commonjs/internal/FilesystemFolder.java (hunk 1 above)
Comment (jpg0) [reply bundles/org.openhab.automation.module.script.graaljs/src/main/java/org/openhab/automation/module/script/graaljs/commonjs/internal/FilesystemFolder.java]: Apologies that this took me so long, for some reason I didn't see this comment. The code was formatted with the default Intellij settings, but I've reformatted now with the default Eclipse settings (looks like the main difference was 4 spaces vs 2 spaces indentation).

**Comment by jpg0** [comment]
> which I see several issues with

Are there any issues that I can begin to work on now?

**Comment by jpg0** [comment]
@martinvw I formatted using the default Eclipse formatter, and it used tabs instead of spaces, which has failed the build. Is there a specific formatter documented or downloadable anywhere, or should I just switch tabs to (4x) spaces?

File: bundles/org.openhab.automation.module.script.graaljs/pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpg0) [reply bundles/org.openhab.automation.module.script.graaljs/pom.xml]: Ok, now on the LTS release.

**Comment by jpg0** [comment]
I have two features that I had added after opening this PR:
- Enforcing single-thread per Graal JS context for rule executions
- Allowing host preset imports via standard JS module system

Currently these are not included; I'm happy to keep going on this PR without them (I'll submit as subsequent ones), but if you'd prefer to include them, let me know.

**Comment by cpmeister** [comment]
2.5.x is our current working branch for openhab-addons. Please change your PR to merge to that branch instead.

**Comment by jpg0** [comment]
TBH I've assumed that this PR is dead as it's been open for 4 months. My code has moved on as well as OH, although I know that it's still compatible as I'm still actively using and maintaining it.

If this PR has a hope of actually being merged I'm happy to update, but if it's going nowhere then I can close it.

**Comment by kaikreuzer** [comment]
@jpg0 Definitely not dead or forgotten!
I'd hope that it can move forward as soon as we have a decision on https://github.com/openhab/openhab-core/issues/1319

**Comment by kaikreuzer** [comment]
@cpmeister This PR is definitely for 3.0 and not 2.5, so having the PR against master is just fine in this case!

**Comment by jpg0** [comment]
@kaikreuzer I'll update this PR once that decision has been made, as it will affect this code too. There are quite a few improvements since this was originally submitted.

**Comment by kaikreuzer** [comment]
Thanks 👍

**Comment by openhab-5iver** [comment]
As I mentioned much much earlier, there are several issues that I see in this PR that I believe should be addressed before it is merged. The openhab-addons and openhab-core maintainers may feel differently.

* GraaljsScriptEngineFactory should extend AbstractScriptEngineFactory (this is now possible after [openhab-core/#1251](https://github.com/openhab/openhab-core/pull/1251)).
* Remove everything other than the ScriptEngineFactory and submit the other functionality in another bundle(s). I'll also be removing the helper libraries from the JythonScriptEngineFactory. I've put a lot of thought into this and it will be cleaner to have bundles for just the ScriptEngineFactories, but even more so, it will be better to have separate bundles for the "other stuff".
* This addon conflicts with `org.openhab.core.automation.module.script.internal.NashornScriptEngineFactory`, which will require changes in OHC to cleanup. In testing this bundle, things got messy when installing and uninstalling, and when transitioning to/from Nashorn. I know this is vague, but my testing was months ago :smile:. In an attempt to resolve the issue at that time, I had split NashornScriptEngineFactory from OHC and created a separate addon for it.
* That's all I can remeber ATM. After [openhab-core/#1319](https://github.com/openhab/openhab-core/issues/1319) is resolved, I plan to do a full review of this PR.

**Comment by jpg0** [comment]
Thanks for letting me know the issues that you found! I'll respond here whilst I've got a second...
1. I see no issue with using AbstractScriptEngineFactory, assuming that the existing functionality can be replicated
2. Happy to split things up; the only thing that I believe should remain with the actual script engine in the module loading. This is because it's so core to the language (e.g. without scripts suck!) and matches other languages which include it (such as Python).
3. I agree; there were some other issues with reloading too (some classes were leaking in to the system classloader). I did some work on this in a later commit (after this PR) and believe that I fixed this (as it would also break transforms, which continue to run on Nashorn).

Anyway, I'll look forward to the full review.

**Comment by kaikreuzer** [comment]
For your replacement PR, it would be great if you could go directly for the namespace suggested in https://github.com/openhab/openhab-core/issues/1319#issuecomment-616790602.

**Comment by jpg0** [comment]
Sure!

**Comment by jpg0** [comment]
Superceded by https://github.com/openhab/openhab-addons/pull/8441

I'll close this and work on that one; there have been more changes and the add-on is now more mature, but requires some changes in code it depends on.

**Pull request closed without merging**